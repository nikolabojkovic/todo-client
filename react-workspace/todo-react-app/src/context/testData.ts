import { 
  ITodo,
  IPaging, 
  ISort, 
  SortDirection, 
  IFilter,
  StateFilter, 
  IGeneralSettings, 
  IPaginationSettings, 
  ISearchSettings, 
  ISettings, 
  ListContainerType, 
  PaginationType, 
  BackgroundColor,
  IThemeSettings,
  BsThemes
} from "../models";
import { DisplayMode, IState } from "./";

export const todoListTestData: ITodo[] = [
  {
    id: 1,
    title: "Task 1",
    description: "Description 1",
    completed: false,
    createdAt: new Date(2022, 1, 4)
  } as ITodo,
  {
    id: 2,
    title: "Task 2",
    description: "Description 2",
    completed: true,
    createdAt: new Date(2022, 1, 5)
  } as ITodo,
  {
    id: 3,
    title: "Task 3",
    description: "Description 3",
    completed: false,
    createdAt: new Date(2022, 1, 6)
  } as ITodo,
  {
    id: 4,
    title: "Task 4",
    description: "Description 4",
    completed: false,
    createdAt: new Date(2022, 1, 7)
  } as ITodo,
  {
    id: 6,
    title: "Task 6",
    description: "Description 6",
    completed: false,
    createdAt: new Date(2022, 2, 4)
  } as ITodo,
  {
    id: 5,
    title: "Task 5",
    description: "Description 5",
    completed: false,
    createdAt: new Date(2022, 3, 8)
  } as ITodo
] as ITodo[];

export const stateTestData: IState = 
{
  originalList: todoListTestData,
  displayList: todoListTestData,
  displayMode: DisplayMode.All,
  search: { searchTerm: '' },
  filter: { state: StateFilter.all } as IFilter,
  sort: { column: 'createdAt', direction: SortDirection.Asc} as ISort,
  paging: {
    totalCount: 6,
    activePage: 1,
    itemsPerPage: 5,
    startIndex: 0,
    endIndex: 5,
  } as IPaging,
  settings: {
    general: {
      isConfirmEnabled: true,
      isPaginationEnabled: true,
      isInfiniteScrollEnabled: false,
      listSizeType: ListContainerType.Dynamic,
      fixedListSize: 200
    } as IGeneralSettings,
    search: {
      isSearchOnKeyPressEnabled: false,
      debounceTime: 500
    } as ISearchSettings,
    pagination: {
      paginationType: PaginationType.Classic,
      maxVisiblePages: 3
    } as IPaginationSettings,
    theme: {
      backgroundColor: BackgroundColor.DarkGray,
      primaryColor: '#ff9900',
      bsTheme: BsThemes.Light,
      primaryColorTopCord: -35,
      primaryColorLefCord: 81
    } as IThemeSettings
  } as ISettings
} as IState;