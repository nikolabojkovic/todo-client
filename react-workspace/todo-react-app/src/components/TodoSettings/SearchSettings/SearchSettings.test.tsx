import renderer, { act } from 'react-test-renderer';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';

import { IState, TodosContext, TodosDispatchContext, stateTestData } from "../../../context";
import { SearchSettings } from "./SearchSettings";
import { IAction, ISearchSettings, ISettings, TodoActions } from '../../../models';

describe('SearchSettings', () => {
  const globalContext = {
    state: {
      ...stateTestData,
    },
    dispatch: jest.fn()
  };
  
  it('match snapshot search settings', () => {
    const context = {
      ...globalContext,
      state: { 
        ...globalContext.state,
        isLoading: false 
      }
    };
    const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
          <SearchSettings></SearchSettings>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
    const tree = renderer.create(
      jsxElement
    ).toJSON();
    render(jsxElement);
    expect(tree).toMatchSnapshot();
  });

  it('match snapshot search on key press settings', () => {
    const context = {
      ...globalContext,
      state: { 
        ...globalContext.state,
        isLoading: false, 
        settings: {
          ...globalContext.state.settings,
          search: {
            ...globalContext.state.settings.search,
            isSearchOnKeyPressEnabled: true
          } as ISearchSettings
        } as ISettings
      } as IState
    };
    const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
          <SearchSettings></SearchSettings>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
    const tree = renderer.create(
      jsxElement
    ).toJSON();
    render(jsxElement);
    expect(tree).toMatchSnapshot();
  });

  describe('actions', () => {
    it('should update settings on search on key press enabled swich toggle', async () => {
      const context = {
        state: {
          ...stateTestData,
          search: { searchTerm: '' },
          settings: { 
            ...stateTestData.settings,
            search: {
              ...stateTestData.settings.search,
              isSearchOnKeyPressEnabled: false
            }
          }
        },
        dispatch: jest.fn()
      };
      const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
            <SearchSettings></SearchSettings>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
      await act( async () => {
        render(jsxElement);
      });
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          search: {
            ...stateTestData.settings.search,
            isSearchOnKeyPressEnabled: true
          } as ISearchSettings
        }
      } as IAction;
  
      const toggle = screen.getByTestId('search-on-keypress-switch');
      fireEvent.click(toggle);
   
      await waitFor(() => {
        expect(context.dispatch).toBeCalledWith(action);
      }); 
    });
    it('should update settings on list size input changed', async () => {
      const context = {
        ...globalContext,
        state: { 
          ...globalContext.state,
          isLoading: false,
          settings: {
            ...globalContext.state.settings,
            search: {
              ...stateTestData.settings.search,
              isSearchOnKeyPressEnabled: true
            }
          } as ISettings
        } as IState
      };
      render(
        (
          <TodosContext.Provider value={context.state}>
            <TodosDispatchContext.Provider value={context.dispatch} >
              <SearchSettings></SearchSettings>
            </TodosDispatchContext.Provider>
          </TodosContext.Provider>
        )
      ); 
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          search: {
            ...context.state.settings.search,
            debounceTime: 600
          } as ISearchSettings
        }
      } as IAction;
        
      const listSizeInput = screen.getByTestId('debounce-time-input');
      fireEvent.change(listSizeInput, {target: {value: 600}});
     
      expect(context.dispatch).toBeCalledWith(action);
    });
  });
});