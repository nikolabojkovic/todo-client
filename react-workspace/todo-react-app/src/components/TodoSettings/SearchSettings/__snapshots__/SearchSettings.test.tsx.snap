// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchSettings match snapshot search on key press settings 1`] = `
<div
  className="App__settings__group"
>
  <label
    className="App__settings__group-label"
  >
    Search
  </label>
  <div
    className="App__settings__group__item"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        Search on key press
      </label>
      <div
        className="ms-auto form-switch"
      >
        <input
          checked={true}
          className="form-check-input"
          data-testid="search-on-keypress-switch"
          disabled={false}
          id="keyPress-switch"
          onChange={[Function]}
          type="checkbox"
        />
      </div>
    </div>
  </div>
  <div
    className="App__settings__group__item ms-5"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        Debounce time (ms)
      </label>
      <input
        className="ms-auto form-control form-control-sm"
        data-testid="debounce-time-input"
        id="debounce-time-input"
        onChange={[Function]}
        placeholder="0"
        type="number"
        value={500}
      />
    </div>
  </div>
</div>
`;

exports[`SearchSettings match snapshot search settings 1`] = `
<div
  className="App__settings__group"
>
  <label
    className="App__settings__group-label"
  >
    Search
  </label>
  <div
    className="App__settings__group__item"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        Search on key press
      </label>
      <div
        className="ms-auto form-switch"
      >
        <input
          checked={false}
          className="form-check-input"
          data-testid="search-on-keypress-switch"
          disabled={false}
          id="keyPress-switch"
          onChange={[Function]}
          type="checkbox"
        />
      </div>
    </div>
  </div>
</div>
`;
